### PR TITLE
[CodeGen][NewPM] Port RegUsageInfoPropagation pass to NPM

### PR DESCRIPTION
**Pull request by optimisan**

### CODE REVIEW

**Comment by optimisan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114010?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#114010** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114010?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#113874** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113874?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113873** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113873?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @optimisan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/CodeGen/RegUsageInfoPropagate.cpp [hunk 1]
```diff
@@ -95,12 +107,29 @@ static const Function *findCalledFunction(const Module &M,
...
+  return RUIP.run(MF);
+}
+
+PreservedAnalyses
+RegUsageInfoPropagationPass::run(MachineFunction &MF,
+                                 MachineFunctionAnalysisManager &MFAM) {
+  Module &MFA = *MF.getFunction().getParent();
+  auto *PRUI = MFAM.getResult<ModuleAnalysisManagerMachineFunctionProxy>(MF)
+                   .getCachedResult<PhysicalRegisterUsageAnalysis>(MFA);
+  assert(PRUI && "PhysicalRegisterUsageAnalysis not available");
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegUsageInfoPropagate.cpp:124]: Why not just use getResult and delete the assert?

Comment (optimisan) [reply llvm/lib/CodeGen/RegUsageInfoPropagate.cpp:124]: It is an outer analysis so cannot run it from an inner IR,

**Review by arsenm — APPROVED** [review]

**Comment by optimisan** [comment]
### Merge activity

* **Nov 14, 8:37 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/114010).

**Pull request merged**